content/blog/en/dojo-middleware.md

```diff
@@ -0,0 +1,213 @@
+---
+title: How to use middleware to create reusable functionality across widgets
+date: 2020-02-03
+author: Anthony Gubler & Rene Rubalcava
+---
+
+Dojo provides a [middleware system](https://dojo.io/learn/middleware/introduction) for developing widgets for your applications. There is a comprehensive list of [available middleware](https://dojo.io/learn/middleware/available-middleware) for managing local widget state, styling, or DOM related information.
+
+<!-- more -->
+
+Middleware are really interesting because they can help you interact with the DOM or with the properties of your widget with retaining a reactive architecture.
+
+Middleware gets created with the same factory used to create a widget, except instead of returning a `vnode`, an object or function gets returned defining the API of the middleware and can get used to do some extra work for your widget.
+
+The Dojo documentation touches on [creating your own middleware](https://dojo.io/learn/middleware/middleware-fundamentals#creating-middleware). How could you implement your own custom middleware for your own widgets?
+
+## Validation Middleware
+
+Maybe we're building form-based widgets and we want to provide our own validation. For example, we might want to validate that a phone number is entered correctly.
+
+In this case, we're interested in wrapping an input in some form of validation. First, we need to create a `PhoneValidator` widget to wrap DOM `input`. The result would look something like this:
+
+> src/widgets/PhoneNumber.tsx
+
+```tsx
+import { create, tsx } from '@dojo/framework/core/vdom';
+import icache from '@dojo/framework/core/middleware/icache';
+
+import PhoneValidator from './PhoneValidator';
+
+import * as css from './styles/PhoneNumber.m.css';
+
+const factory = create({ icache });
+
+export const PhoneNumber = factory(function PhoneNumber({ middleware: { icache } }) {
+	// use local phone value to pass to validator
+	const phone = icache.getOrSet('phone', '');
+	return (
+		<PhoneValidator phone={phone}>
+			<input
+				placeholder="Enter phone number"
+				pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"
+				required
+				classes={[css.root]}
+				type="tel"
+				onkeyup={(e) => {
+					icache.set('phone', (e.target as HTMLInputElement).value);
+				}}
+			/>
+		</PhoneValidator>
+	);
+});
+
+export default PhoneNumber;
+```
+
+Here we want the `PhoneValidator` to place a red or green outline to the input to inform the user if it's a valid phone number or not. It's pretty simple, but is something that could be reused across multiple applications.
+
+> src/widgets/PhoneValidator.tsx
+
+```tsx
+import { create, tsx } from '@dojo/framework/core/vdom';
+import phoneNumberMiddleware from '../middleware/phoneNumberMiddleware';
```

Comment:
can this just be called `phoneNumber`? We tend not to use a suffix to indicate middleware elsewhere